### ISSUE TITLE
Official Technical IWXXM specification?

### ISSUE THREAD

**Issue by ilkkarinne** [wmo-im/iwxxm#89]
I'm involved in a software project for converting various aviation messages between TAC, IWXXM, Java and JSON objects. Is there a definitive or at least authoritative technical guidance / specification on using IWXXM 2.1 encoding for official WMO / ICAO aviation weather messages?

The schemas, examples, schematron rules and the brief TAC-to-XML guidance are good, but there are lots of things left unspecified, such as the use of code (required?) lists. 

IMHO there must be a document with at least similar level of technical rules as ICAO Annex 3 provides for TAC to ensure an acceptable level of interoperability between the various IWXXM software implementations.

**Comment by blchoy** [comment]
Thank you for your suggestion.  As you may know, Annex 3 is currently being restructured (see Section 3.3 of https://www.icao.int/airnavigation/METP/Panel%20Meetings/METP2%20Final%20report.pdf) and that should better align with IWXXM than TAC.  At this moment, if you have any IWXXM examples which best describe an instance of TAC and would like to share with the community, you are most welcome to put it on https://github.com/wmo-im/iwxxm-translation

**Comment by braeckel** [comment]
Another way to think about this question would be to consider IWXXM the technical specification and to define translations to JSON and TAC.  For JSON there are existing, general-purpose XSLT scripts (for example, [this](https://github.com/bramstein/xsltjson), among many others) which can translate any XML message to JSON and could theoretically be used with IWXXM, FIXM, and AIXM.  This would simplify the problem by making the IWXXM schemas the authoritative definition, which would eliminate a dimension of the problem by fixing the relationship between IWXXM and other formats.  If we (WMO) identify a mechanism for translating from XML to JSON it may be the most straightforward and least ambiguous way of translating the well-known XML requirements into JSON requirements - human-readable documentation in PANS-MET or an ICAO Annex may be more subject to interpretation.  The PANS-MET process is also likely to take a few years to be completed.

The TAC question is significantly more challenging.  We've previously discussed internally whether WMO should publish authoritative XSLT scripts to translate from IWXXM to TAC.  For IWXXM 1.0 I developed some unreleased early/draft scripts to do this for METAR, SPECI, TAF, and part of SIGMET, but these have not been updated for some time and would need to be significantly extended and tested to fully meet TAC translation requirements.  Other organizations, including commercial organizations, have been sharing links to their TAC/IWXXM translation tools and it may be worthwhile to compare your tool against the results of these tools.  

In addition, we've created a [repository](https://github.com/wmo-im/iwxxm-translation) of IWXXM translation examples (IWXXM/TAC pairs that are equivalent) as a testing suite for verifying translation software.  This currently only has the "stock" IWXXM examples but the intent is for any and all organizations to contribute to this set and add more exotic cases.  Once enough TAC/IWXXM pairs have been added to this repository it could become a useful tool for comparing translation software and approaches.  I encourage you to contribute your interesting cases to this repository (create a Pull Request), but note that there are many, many TAC cases that are not ICAO compliant or incorrectly formatted and can not be represented in IWXXM.  We are not trying to address the problem of TAC data quality more generally, only the cases of ICAO-compliant TAC messages and their equivalent IWXXM messages.
